File: src/lib/api/updates.ts
import type { UpdateStatus } from "@/features/updater/types"
import { apiGet, apiPost } from "./client"

/**
 * Self-update, driven by the Rust side and reported here.
 *
 * Reading and checking need a session; downloading and installing are admin
 * routes, because they close and replace the program running on the till. All
 * three writes answer with the status they moved to, so a caller can seed the
 * cache instead of waiting for the next poll.
 */

export function getUpdateStatus(): Promise<UpdateStatus> {
  return apiGet<UpdateStatus>("/updates")
}

/** Resolves once the check has finished, with the phase it ended in. */
export function checkForUpdate(): Promise<UpdateStatus> {
  return apiPost<UpdateStatus>("/updates/check")
}

/** Returns as soon as the download is running; progress arrives via the status. */
export function downloadUpdate(): Promise<UpdateStatus> {
  return apiPost<UpdateStatus>("/updates/download")
}

/** The app exits shortly after this answers; the installer relaunches it. */
export function installUpdate(): Promise<UpdateStatus> {
  return apiPost<UpdateStatus>("/updates/install")
}

File: src-tauri/src/http/routes/updates.rs
//! Self-update: status, check, download, install.
//!
//! Thin on purpose — the state machine is [`crate::updater::Updater`]. Reading
//! the status and asking for a check need only a session, because knowing a
//! newer version exists helps a cashier tell the owner. Downloading and
//! installing close and replace the program on the till, so they are admin
//! actions like restoring a backup.
//!
//! All three writes answer `202 Accepted` with the status they moved to: the
//! download and the install carry on after the response, and `check` returns
//! the phase it ended in.

use axum::extract::State;
use axum::http::StatusCode;
use axum::routing::{get, post};
use axum::Router;

use crate::http::error::ApiResult;
use crate::http::AppState;
use crate::updater::UpdateStatus;
use crate::utils::AppError;

pub fn session() -> Router<AppState> {
    Router::new()
        .route("/updates", get(status))
        .route("/updates/check", post(check))
}

pub fn admin() -> Router<AppState> {
    Router::new()
        .route("/updates/download", post(download))
        .route("/updates/install", post(install))
}

async fn status(State(state): State<AppState>) -> axum::Json<UpdateStatus> {
    axum::Json(state.updater.status())
}

type Accepted = ApiResult<(StatusCode, axum::Json<UpdateStatus>)>;

fn accepted(moved_to: Result<UpdateStatus, AppError>) -> Accepted {
    Ok((StatusCode::ACCEPTED, axum::Json(moved_to?)))
}

async fn check(State(state): State<AppState>) -> Accepted {
    accepted(state.updater.check().await)
}

async fn download(State(state): State<AppState>) -> Accepted {
    accepted(state.updater.download())
}

async fn install(State(state): State<AppState>) -> Accepted {
    accepted(state.updater.install())
}
